src/libmongoc/tests/unified/runner.c

```diff
@@ -1256,26 +1256,33 @@ test_run_distinct_workaround (test_t *test, bson_error_t *error)
    /* Get the database/collection name from each collection entity. */
    BSON_FOREACH (test->test_file->create_entities, iter)
    {
-      bson_t create_entity;
       bson_t entity_bson;
       char *coll_name = NULL;
       char *db_id = NULL;
       char *db_name = NULL;
       mongoc_database_t *db = NULL;
       bson_iter_t entity_iter;
 
-      bson_iter_bson (&iter, &create_entity);
-      if (!bson_has_field (&create_entity, "collection")) {
-         continue;
-      }
       if (!BSON_ITER_HOLDS_DOCUMENT (&iter)) {
          test_set_error (error,
                          "unexpected non-document createEntity: %s",
-                         tmp_json (&create_entity));
+                         bson_iter_key (&iter));
          goto done;
       }
 
       bson_iter_recurse (&iter, &entity_iter);
+
+      if (!bson_iter_find (&entity_iter, "collection")) {
+         continue;
+      }
+
+      if (!BSON_ITER_HOLDS_DOCUMENT (&entity_iter)) {
+         test_set_error (error,
+                         "unexpected non-document in iter: %s",
+                         bson_iter_key (&entity_iter));
+         goto done;
+      }
```

Comment:
Looking at the unified test format tests, there is no test using `distinct`, so I believe that this wasn't run until the Versioned API tests came in. It was the second `bson_iter_bson` that tripped, and I suspect that this was due to the iterator not being where we expect it to be after `bson_iter_recurse`, but I'm not entirely sure since I don't know how `bson_iter_t` works internally.

Reply:
Ah ok, surprised `distinct` never appeared.

Reply:
@alcaeus's suspicion is right. `bson_iter_recurse` initializes the child `bson_iter_t` with the document / array iterated.

On initialization with `bson_iter_init`, the `bson_iter_t` sets a zero offset. The `bson_iter_t` is not pointing to the first key/value pair until the first call to `bson_iter_next` (or something that calls the next function, like `bson_iter_find`).

The issue was a missing call to `bson_iter_find (&entity_iter, "collection");` after the second call to `bson_iter_recurse`.

This works just as well!